paddle/fluid/framework/operator.cc

```diff
@@ -1180,12 +1178,38 @@ Scope* OperatorWithKernel::PrepareData(
         pre_scope_ = nullptr;
       }
 
-      auto* trans_var = new_scope->Var(var_name);
-      input_vars[i] = trans_var;
-
-      Tensor out;
-      TransformData(expected_kernel_key, kernel_type_for_var, *tensor_in, &out);
-      SetTensorToVariable(*var, out, trans_var);
+      if (no_buffer_ins && no_buffer_ins->count(var_name_item.first) > 0) {
```

Comment:
You cannot move `if (no_buffer_ins && no_buffer_ins->count(var_name_item.first) > 0)` here. 

In the design of `no_buffer_ins`, the tensor buffer of vars (i.e. the `holder_` field of Tensor) inside `no_buffer_ins` may be written (clear or written with other value) by other threads. Therefore, it is not thread safe if you read or write the tensor buffer of vars inside `no_buffer_vars`. Because the data transform does not handle these vars, the current implementation of Paddle is right when running ops in multi-threads. The only 2 methods you can use is `Tensor::dims()` and `Tensor::lod()`. Any Tensor methods which read or write `Tensor::holder_` is not allowed.

 In the current implementation of this PR, you check whether the tensor is initialized , which may cause error in multi-threads.
https://github.com/PaddlePaddle/Paddle/blob/a400a2f37fc4d4c636171be66088641b3889d906/paddle/fluid/framework/operator.cc#L1121

Reply:
Ok, So If I understand correctly when there is no_buffer_ins we should not call Tensor methods you listed. I have updated code (only operator.cc) Please tell me if this is fine approach.